torch/_decomp/decompositions.py

```diff
@@ -268,6 +307,49 @@ def log_sigmoid_backward(grad_output: Tensor, self: Tensor, buffer: Tensor) -> T
     # return (max_deriv - sign * (buffer / (1 + buffer))) * grad_output
 
 
+def apply_loss_reduction(loss: Tensor, reduction: int):
+    if reduction == Reduction.MEAN.value:
+        return torch.mean(loss)
+    elif reduction == Reduction.SUM.value:
+        return torch.sum(loss)
+    else:
+        return loss
+
+
+def to_real_dtype(dtype: torch.dtype):
+    if dtype == torch.complex32:
+        return torch.float16
+    elif dtype == torch.complex64:
+        return torch.float32
+    elif dtype == torch.complex128:
+        return torch.float64
+
+
+@register_decomposition(aten.l1_loss)
+def l1_loss(self: Tensor, target: Tensor, reduction: int = Reduction.MEAN.value) -> Tensor:
+    loss = (self - target).abs()
+    # PyTorch semantics result in the output of l1_loss having the corresponding
+    # real dtype to self.  This may not happen without explicit casting if say
+    # self: complex64 and target: float64, which results in loss: float64
```

Comment:
that's weird though, abs() is supposed to take complex to real so shouldn't this happen automatically

Reply:
Oh, it does, the problem is that self can be complex64, and then target is float64, so after we do (loss - target) we end up with a complex128 value. Then, calling abs on that gives us a float64, which is not the corresponding real type for complex64.

Reply:
By the way, if it makes you feel better, our actual kernel does something similar :P 

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L521